src/stm32/i2c.c

```diff
@@ -51,18 +51,70 @@ static const struct i2c_info i2c_bus[] = {
 #endif
 };
 
+static void
+i2c_us_delay(uint32_t us)
+{
+    uint32_t timeout = timer_read_time() + timer_from_us(us);
+    while (timer_is_before(timer_read_time(), timeout))
+      ;
+}
+
+static void
+i2c_init(I2C_TypeDef *i2c)
+{
+    uint32_t pclk = get_pclock_frequency((uint32_t)i2c);
+    i2c->CR2 = pclk / 1000000;
+    i2c->CCR = pclk / 100000 / 2;
+    i2c->TRISE = (pclk / 1000000) + 1;
+    i2c->CR1 = I2C_CR1_PE;
+}
+
 // Work around stm32 errata causing busy bit to be stuck
+// "2.9.7 I2C analog filter may provide wrong value, locking BUSY
+// flag and preventing master mode entry"
 static void
-i2c_busy_errata(uint8_t scl_pin, uint8_t sda_pin)
+i2c_stm32f1_busy_errata(I2C_TypeDef *i2c)
 {
-    if (! CONFIG_MACH_STM32F1)
+    if (!CONFIG_MACH_STM32F1)
         return;
-    gpio_peripheral(scl_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, 1);
-    gpio_peripheral(sda_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, 1);
-    gpio_peripheral(sda_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, -1);
-    gpio_peripheral(scl_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, -1);
-    gpio_peripheral(scl_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, 1);
-    gpio_peripheral(sda_pin, GPIO_OUTPUT | GPIO_OPEN_DRAIN, 1);
+
+    const struct i2c_info *ii =
+      container_of((I2C_TypeDef * const *)i2c, struct i2c_info, i2c);
+
+    i2c->SR1;
+    i2c->SR2;
+    i2c->DR;
+    i2c->CR1 = 0;
+    i2c->CR2 = 0;
+    i2c->DR = 0;
+
+    GPIO_TypeDef *scl_regs = digital_regs[GPIO2PORT(ii->scl_pin)];
+    uint32_t scl_bit = GPIO2BIT(ii->scl_pin);
+    GPIO_TypeDef *sda_regs = digital_regs[GPIO2PORT(ii->sda_pin)];
+    uint32_t sda_bit = GPIO2BIT(ii->sda_pin);
+    #define SCL_ODR_SET(v) do { scl_regs->BSRR = (v) ? (scl_bit<<16) : scl_bit; } while (0)
```

Comment:
Seems to be reversed.
Generally, here is gpio_write()

```
// src/stm32/gpio.c
void
gpio_out_write(struct gpio_out g, uint32_t val)
{
    GPIO_TypeDef *regs = g.regs;
    if (val)
        regs->BSRR = g.bit;
    else
        regs->BSRR = g.bit << 16;
}
```

Reply:
Ugh, quite. (Thank you for the in-depth review here!) Is there some way that I can just use `gpio_out_write`? Maybe I should synthesize a `struct gpio_out` for the scl/sda pins?